### DISCUSSION TITLE
WinUI implementation is objectivly broken

### DISCUSSION THREAD

**Discussion by axa88** [dotnet-bluetooth-le/dotnet-bluetooth-le#876] [Q&A] [answered]
After filing and observing other critical issues ie #850, #875, it appears that WinUI has high malfunction.
Curious if there is any road map to address?

**Comment by janusw** [comment]
The Windows implementation is the youngest of all the four platforms we support. If it does not work well for you, maybe you can help to improve it? I'm not natively developing on Windows, so it's quite hard for me to reproduce and debug any Windows-related issues.

**Reply by axa88** [reply]
Unfortunately I don't have the ability either...

**Reply by AskBojesen** [reply]
Can you create a simple WINUI example that shows it it is broken?
I use the plugin extensively in Windows 11 , .Net 8 C# without any errors

**Reply by axa88** [reply]
Ok, I should certainly take the opportunity to demonstrate this as youre offering to look into it.  I'm confident the are serious issues making it unusable. I'll be in touch

**Reply by axa88** [reply]
@AskBojesen 
Inversely, might you have, or be able to point me to, something that works for you? I'm assuming of course you're saying the windows implementation is working well for you? 

There are at least a few other reports indicating the implementation is malfunctioning , but before I could get something 'simple' together I'd like to investigate something that reportedly works.

**Reply by AskBojesen** [reply]
I have created the WinConsole demo project, in which we can test different scenarios - also the disconnect issue, which I do not have a fix for btw

**Reply by axa88** [reply]
> I have created the WinConsole demo project, in which we can test different scenarios - also the disconnect issue, which I do not have a fix for btw

alright ill try this demo and see if my findings line up.

**Comment by axa88** [comment]
@AskBojesen 
Giving the console demo a try it basically confirms the issues that i am experiencing.

Where its possible after just a couple of connect/disconnect cycles, Disconnection will fail, and further use of the API is not possible.
Rational: This in and of itself is basically what makes me determine the API 'objectionably broken'.

Bonding doesn't work.  Its funny cuz i recall initially testing I found it did.  But no matter what i do API Bonding will fail when outside of this API it seems to work fine.
This was confirmed for me when using this demo's naive Pairing process, which does result in a Bond. (Im not sure we are using Pair/Bond terms correctly but i digress).  Oddly Ive never reported this as a bug as im confused why this is happening.

Rational: On Android security requirements often require a Bond initiated by the application, so i would hope to be able to carry out the same process from the windows application without need for special programming/logic... Perhaps iOS has the same requirement but I've never tried it.

**Reply by axa88** [reply]
@AskBojesen 

I stepped through what the API and demo are both doing natively, and i see the difference and hence my issue.

The API calls Pairing while the demo using `custom` pairing and passes the parameter `DevicePairingKinds.ConfirmOnly`.
Having created my peripheral device to _not_ use any pairing interaction, i didn't assume this was an issue.
Anyway Im glad this has helped me figure out that issue, but im curious how people are using Bonding without any Pairing option support? Is anyone using Bonding?

**Reply by axa88** [reply]
Bonding works now #906.
Perhaps I'll look into the connection issue next 🤔

**Comment by axa88** [comment]
Ive rewritten the Windows implementation and have a POC that confirms it works flawlessly.
Extensive changes required, along with full support for bond/pairing, #906, I think major changes to this API should be entertained.
Perhaps i'll open a discussion on it.

**Reply by Bainion** [reply]
You've made a fork with working Windows implementation? Could you share it? I also have problem with disconnect that sometimes hangs and if you use CTS, it breaks the scanning. Also sometimes the steps connect -> service discovery -> characteristic discovery takes longer than necessary (or stays stuck) and can break the whole connection procedure.

**Reply by axa88** [reply]
@Bainion
Sure, but you're probably not going to like it.

iirc, this is the last branch that I merged with the main project before i started a new one entirely.

`https://github.com/axa88/dotnet-bluetooth-le/tree/refactor/ModernClean`

If you just want to get windows working correctly you can focus on extracting the changes in the windows Adapter, iirc it implements a new device manager object and that's what implements the windows api correctly, imho...

iirc i made the minimal updates to the windows console test project so it can be tested by what's already available. but im not taking credit for any of that, it's rather messy.

Bonus. it also fully supports all modes of device pairing besides out of band.

good luck

**Reply by Bainion** [reply]
Thanks for the link

I'm sorry to be a bother again but I tried to build it but I can't seem to make it work. I tried both with cake and VS2022.
I'm only trying to build the Plugin.BLE for the dll and it tells me this with VS2022:

```Build started at 16:09...
1>------ Build started: Project: Plugin.BLE, Configuration: Release Any CPU ------
1>Plugin.BLE -> D:\Projects\Bluetooth\dotnet-bluetooth-le\Source\Plugin.BLE\bin\Release\netstandard2.0\Plugin.BLE.dll
1>Plugin.BLE -> D:\Projects\Bluetooth\dotnet-bluetooth-le\Source\Plugin.BLE\bin\Release\net8.0-android34.0\Plugin.BLE.dll
1>C:\Program Files\dotnet\sdk\9.0.308\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.EolTargetFrameworks.targets(38,5): error NETSDK1202: The workload 'net8.0-ios' is out of support and will not receive security updates in the future. Please refer to https://aka.ms/maui-support-policy for more information about the support policy.
1>C:\Program Files\dotnet\sdk\9.0.308\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.TargetFrameworkInference.targets(230,5): error NETSDK1135: SupportedOSPlatformVersion 10.0 cannot be higher than TargetPlatformVersion 1.0.
1>Done building project "Plugin.BLE.csproj" -- FAILED.
1>C:\Program Files\dotnet\sdk\9.0.308\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.EolTargetFrameworks.targets(38,5): error NETSDK1202: The workload 'net8.0-maccatalyst' is out of support and will not receive security updates in the future. Please refer to https://aka.ms/maui-support-policy for more information about the support policy.
1>C:\Program Files\dotnet\sdk\9.0.308\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.TargetFrameworkInference.targets(230,5): error NETSDK1135: SupportedOSPlatformVersion 13.1 cannot be higher than TargetPlatformVersion 1.0.
1>Done building project "Plugin.BLE.csproj" -- FAILED.
1>Plugin.BLE -> D:\Projects\Bluetooth\dotnet-bluetooth-le\Source\Plugin.BLE\bin\Release\monoandroid13.0\Plugin.BLE.dll
1>Plugin.BLE -> D:\Projects\Bluetooth\dotnet-bluetooth-le\Source\Plugin.BLE\bin\Release\xamarin.ios10\Plugin.BLE.dll
1>Plugin.BLE -> D:\Projects\Bluetooth\dotnet-bluetooth-le\Source\Plugin.BLE\bin\Release\xamarin.mac20\Plugin.BLE.dll
1>Plugin.BLE -> D:\Projects\Bluetooth\dotnet-bluetooth-le\Source\Plugin.BLE\bin\Release\uap10.0.19041\Plugin.BLE.dll
1>Plugin.BLE -> D:\Projects\Bluetooth\dotnet-bluetooth-le\Source\Plugin.BLE\bin\Release\net8.0-windows10.0.19041\Plugin.BLE.dll
1>Plugin.BLE -> D:\Projects\Bluetooth\dotnet-bluetooth-le\Source\Plugin.BLE\bin\Release\net8.0-windows10.0.22000.0\Plugin.BLE.dll
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 16:09 and took 01.432 seconds ==========
```

With cake it tells me it can't download the addins Cake.Git (first line of the cake script) and if I comment the function using the addins, it also blocks for reason of end of support (on the 'Restore' step).

Did you also have such problem or do you have any idea of how to fix it?
If not, could you maybe provide at least the Windows (.NET8) DLLs (as the other seem to work) please?

**Reply by axa88** [reply]
I really dont know whats the issue is there, try upgrading your targets to .net9. ive gone straight to v10 for notable UI performance improvements.
Sorry, If i get the time someday ill release an entirely new lib for windows and android.  But otherwise, as mentioned you can cherry pick to get windows working. Even if dated, the old library seems to work fine on other platforms.

**Reply by Bainion** [reply]
I upgraded the targets and I managed to build it on VS2022. Thanks for the help.
I hope a fix for the Windows implementation will be done in the main project in the future.

**Reply by axa88** [reply]
> I hope a fix for the Windows implementation will be done in the main project in the future.

ikr. give that version a try, if/when you see the improvement maybe others might become interested.

**Comment by axa88** [comment, marked as answer]
And it seems like it's going to stay this way indefinitely

**Comment by sinatrocious** [comment]
The library works great on iOS and Android. But it's totally NOT working on windows.

Windows symptoms (compared to other platforms):
- disconnecting is MUCH slower;
- errors ratio (e.g. null when requesting service, exceptions during disconnecting, exceptions when sending, etc.) is MUCH higher;
- online and working device disappea from the scan results without reason;
- online and working device doesn't appear in scan ever (very likely to happen after error of any kind);
- restarting app often doesn't fix problems, need restart of PC and device

In other words, starting windows app, seeing say 20 BLE devices.. scanning after 10 minutes - 0 devices. Where are all those devices gone? Eaten by library?
